File: Majority Element.py
'''

LeetCode 169. Majority Element
https://leetcode.com/problems/majority-element/

Using the Boyer–Moore majority vote algorithm
https://en.wikipedia.org/wiki/Boyer%E2%80%93Moore_majority_vote_algorithm

'''
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        m = 0
        i = 0
        
        for n in nums:
            if i == 0:
                m = n
                i += 1
            elif m == n:
                i += 1
            else:
                i -= 1
        
        return m

File: MajorityElement.c
/*
    LeetCode 169. Majority Element
    Using Boyer–Moore majority vote algorithm
*/

int majorityElement(int* nums, int numsSize){
    int ans = nums[0];
    unsigned int ctr = 1,
                 ptr = 1;
    
    while(ptr<numsSize){
        if (!ctr){
            ans = nums[ptr++];
            ++ctr;
            continue;
        }
        if(ans == nums[ptr++])
            ++ctr;
        else
            --ctr;
    }
    
    return ans;

}